[2단계 - 자동차 경주]세오(진솔) 미션 제출합니다.

jin123457 opened the pull request:
안녕하세요. 케빈😊
주말 잘 보내고 계신가요?ㅎㅎ
1단계에서 피드백 주신 사항과 공통 리팩터링 요구 사항을 반영해 리팩터링을 진행해보았습니다!

# 🎯 리팩터링 후 변경 사항

### 1. controller 역할 분리  [🔗 `cd6b534`](https://github.com/jin123457/javascript-racingcar/commit/cd6b53470a1a99bda37c5453af56a3d23187414b)
기존 `controller`를 `Race`로 변경을 하고 `RaceSetup`, `Winners` 총 3개로 나누어 보았습니다. 😤

* `RaceSetup` 클래스의 역할
  * 사용자에게 자동차 이름을 입력 받고 자동차 리스트를 생성 / 반환
  * 사용자에게 경주 횟수 입력 / 반환
* `Winners` 클래스의 역할
  * 제일 멀리간 자동차를 판단 후 제일 멀리간 자동차의 이름 저장 / 반환

⬆️ 위 두 가지로 케빈이 피드백 주신 controller에 역할 나누기를 반영해보았습니다..!😤

### 2. Car의 moveForward 함수  [🔗 `a817f17`](https://github.com/jin123457/javascript-racingcar/commit/a817f17e910a2a6d73a5f75d95646efb97dcf736)
**변경 전**
``` js
move(randomNumber) {
  if (randomNumber >= MIN.MOVE_CONDITION) {
    this.#position += MOVE_UNIT;
  }
}
```

**변경 후**
``` js
// Race.js
carList.forEach((car) => {
  const moveCondition = randomNumber();
  if (moveCondition >= MIN.MOVE_CONDITION) {
    car.moveForward();
  }
});

// Car.js
moveForward() {
  this.#position += MOVE_UNIT;
}
```
move에서 판단을 하고 전진한다. 라는 것이 '함수의 단일 책임에 어긋나지 않나?'라는 생각이 들어 자동차의 전진하는 조건을 Car에서 하는 것이 아닌 Race에서 판단을 하고 조건에 부합하는 경우 moveFoward 함수를 호출하고 무조건 전진하는 방식으로 수정을 해보았습니다.
케빈은 어떻게 생각하시는 궁금합니다..!🤔

### 3. 네이밍
**EX]**
``` js
class Race {
  startRace() {
    ~~~~
  }
}
```
⬆️ 위와 같이 클래스명에서 Race라는 이름이 있는데 함수에도 race라는 이름을 꼭 넣어줘야 할까? 라는 생각에 race라는 것을 제거해 보았습니다.

# 🙋‍♂️ 이외에 케빈에게 질문하고 싶은 점
* `domain 로직`과 `UI 로직`이 잘 구분되어 있나요?
* utils 폴더에 제가 생각한 도메인 로직과 관련이 애매하고 범용성있는 코드를 모아두었습니다. 
하지만, 제대로 한게 맞을까? 라는 생각이 자꾸 들어 케빈은 폴더를 나눈 것에 대해 어떻게 생각하시는지 궁금합니다.
* 이번 공통 피드백으로 받은 `domain/` 하위의 모듈은 `view/` 하위의 모듈을 의존하지 않아야 한다. 라는 것을 제가 잘 지키고 있는지 궁금합니다.
* 이번 공통 피트백으로 리뷰이가 갖춰야 할 자세에 대해 배웠습니다. (1단계 PR 메시지가 너무 형편없었습니다... 죄송합니다😭)
  이번 PR 메시지는 케빈이 읽기에 편하신지 혹은 더 수정을 했으면 좋겠다라는 것이 있을까요? 혹시 있으시다면 감사하겠습니다.
* 마지막으로 케빈은 개발을 할 때 어느 부분이 제일 설레고 재밌으신가요? 🤔 

1단계에서 피드백을 너무 자세하게 잘 써주셔서 많은 공부가 되고 있습니다. 감사합니다!!😊

jin123457 committed:
Docs : README 요구사항 재정리

[수정된 파일]
README.md : 기능 / 작업 목록을 다시 한 번 정리하며 제가 생각한 기준으로 도메인 로직과 ui로직을 분리하였습니다. 또한 예외처리 사항을 정리 한 후 예외 처리할 사항을 몇 가지 추가했습니다.

jin123457 committed:
Docs : 리팩토리 요구사항 목록 추가

[수정된 파일]
README.md : 2단계에 필요한 리팩토리 요구사항 목록을 체크 목록으로 추가하였습니다.

jin123457 committed:
Rename : model -> domain으로 Car.js 이동

[수정된 파일]
Car.js : model -> domain으로 파일을 이동했습니다.

jin123457 committed:
Rename : 입출력 관련된 로직 view 하위로 이동

[수정된 파일]
errorHandler.js : 입력 후 오류가 있을 경우 에러 메시지를 보여주는 파일이므로 UI로직이라고 판단했습니다.
readLine.js : 입력을 받는 파일이므로 UI로직이라고 판단했습니다.
tryInput.js : 입력 후 오류가 있을 경우 에러 메시지를 보여주고 재입력을 받는 파일이므로 UI로직이라고 판단했습니다.

jin123457 committed:
Rename : 도메인 로직 domain 하위로 이동

[수정된 파일]
positionToStr.js : 자동차의 현재 값을 파라미터로  받고 문자열로 변환해주는 파일이기에 도메인 로직이라고 판단했습니다.
random.js : 무작위 값을 구하는 파일이기에 도메인 로직이라고 판단했습니다.
validation.js : 입력을 받은 값으로 검증하는 것이긴 하지만, 파라미터로 받은 값을 판단하는 것이기에 도메인 로직이라고 판단했습니다.

jin123457 committed:
Rename : controller/ ->  domain/ 하위로 Controller.js 이동

[수정된 파일]
Controller.js : controller.js에 있는 파일의 함수들이 도메인 로직이라고 판단하여 controller/ ->  domain/ 하위로 이동했습니다.

jin123457 committed:
Docs : 리팩토리 요구 사항 목록 현황 반영

[수정된 파일]
README.md : 리팩토리 요구사항 목록에서 현재 처리한 사항들을 반영하였습니다.

jin123457 committed:
Chore : 파일 이동에 맞게 호출 경로 수정

[수정된 파일명]
App.js
- Controller/ -> domain/

Controller.js
- model/ -> domain/
- util/ -> ./

Input.js
- util/ -> ./

Output.js
- util/ -> domain/

jin123457 committed:
Chore : 테스트 코드 파일에서 호출 경로 수정

[수정된 파일]
Car.test.js
- model -> domain/

Controller.test.js
- controller/ -> domain/
- model/ -> domain/

validation.test.js
- util/ -> domain/

jin123457 committed:
Rename : controller -> Race로 파일명  수정

[수정된 파일]
Race.js : controller(기존)에서 Race(신규)로 좀 더 명확하게 파일명을 수정했습니다.

jin123457 committed:
Rename : Race 파일명 수정에 따른 파일명 수정 및 호출 경로 수정

[수정된 파일]
Race.test.js : Race.js 파일명이 수정되면서 테스트하는 파일명 수정 및 호출경로를 수정했습니다.

jin123457 committed:
Chore : prettier 기초 세팅

[수정된 파일]
.prettierrc : 협업을 위한 코드 스타일 기초 세팅

jin123457 committed:
Rename : Race.js 파일명 수정에 따른 호출 경로 수정

[수정된 파일명]
App.js : Race.js 파일명 수정에 따른 호출 경로 수정하였습니다.

jin123457 committed:
Refactor : Race 클래스 역할 분리

[수정된 파일]
Race.js : Race 객체가 많은 역할을 하고 있는 피드백을 반영하고자 Winner, RaceSetup 클래스를 생성하여 Race 클래스에 역할을 경주 역할에 맞는 것들만 남겼습니다.

[생성된 파일]
RaceSetup.js : RaceSetup 클래스는 경주를 시작하기 전 사용자에게 입력을 받고 반환하는 역할을 가지고 있습니다. 역할은 4가지가 있습니다.
- 자동차의 이름을 받은 후 자동차 리스트 생성하기
- 경주 횟수를 입력 받기
- 자동차 리스트를 호출할 시 반환하기
- 경주 횟수 반환하기

Winners.js : Winners 클래스는 경주가 완료된 후 우승자를 판단 후, 우승자의 이름을 반환하는 역할을 가지고 있습니다. 역할은 2가지가 있습니다.
- 제일 멀리간 자동차를 찾고 자동차의 이름들을 저장하기
- 우승자들의 이름을 호출할 시 반환하기

jin123457 committed:
Refactor : winner 함수 분리 구현

[수정된 파일]
Race.js : start 함수에서 winner를 바로 구현하지 않고 따로 printWinners 함수를 빼두어 좀 더 명확하게 표현하기 위해 분리하였습니다.

jin123457 committed:
Chore : OutputView 가능한 받은 값 출력하게끔 수정

[수정된 파일]
Output.js : 기존은 printWinners 함수에서 winners를 파라미터로 받아 join을 했습니다. 하지만 파라미터로 join한 값이 넘어와서 그 상태로 보여주는 것이 나중에 유지보수하기에 편할 것 같다는 생각을 하여 수정하게 되었습니다.

Winners.js : getNames 함수 호출 시 join을 한 뒤에 반환하게끔 수정하였습니다.

jin123457 committed:
Refactor : 자동차 전진 함수 조건 호출 수정

[수정된 파일]
Car.js : 기존 move 함수에서 전진 여부를 판별하고 조건에 부합하면 전진을 했었습니다.
그러므로 move라는 함수에 두가지 책임이 있다고 판단을 했고 단일 책임을 위해 move는 전진하는 기능만 있고 함수의 이름을  move -> moveForward로 수정했습니다.
이렇게 수정을 하므로 자동차의 전진 테스트를 명확하게 할 수 있을 것 같습니다.
Race.js : 랜덤 숫자를 받아와 car 클래스에 move 함수를 무조건 호출하는 것이 아닌 조건적으로 호출을 하는 방식으로 수정하였습니다.

jin123457 committed:
Feat : 자동차 최대 개수 제한 및 경기 최대 라운드 제한 기능 구현

[수정된 파일]
validation.js
- 자동차의 최대 개수를 60개로 제한하는 기능 추가
- 경주 횟수를 최대 MAX_SAFE_INTEGER까지로 제한 기능 추가했습니다.

error.js : RACE_COUNT_OVER_MAX_SAFE_INTEGER 변수 추가했습니다.

constant.js : MAX의 객체 안 CAR_COUNT를 60으로 선언했습니다.

jin123457 committed:
Docs : 기능 구현 목록 및 예외 처리 현황 반영

[수정된 파일]
README.md : 기능 구현 목록 현황 반영을 하였습니다.

jin123457 committed:
Test : Car 클래스 테스트 내용 수정

[수정된 파일]
Car.test.js : Car 클래스에서 move 함수명 변경 및 move 함수 안에 랜덤 숫자 판단 기준 삭제에 따른 테스트 코드 변경을 하였습니다.

jin123457 committed:
Test : Winners 클래스 생성에 따른 테스트 코드 추가

[생성된 파일]
Winners.test.js : Winners 클래스 생성에 따라 단위 테스트 생성 UI 로직에 대한 테스트는 제외하고 도메인 관련된 로직만 테스트를 진행하였습니다.

jin123457 committed:
Test : 예외처리 사항이 추가되어 테스트 코드 기능 추가 구현

[수정된 파일]
validation.test.js : 예외처리 사항이 추가됨으로 테스트 기능 추가적으로 구현하였습니다.
⬇️ 구현한 목록은 아래와 같습니다.
- 자동차는 최대 60대까지 참가 가능하다.
- 자동차의 이름은 1자이상 5자 이하여야 한다.
- 자동차의 이름은 중복되면 안된다.
- 경주 시도의 최대 횟수는 Number.MAX_SAFE_INTEGER 숫자보다 클 수 없다.
error.js : 자동차의 이름이 중복될 경우 사용자에게 보여주는 문구 추가를 하였습니다.
validation.js : 이름이 중복되면 에러 문구를 띄우는 함수 추가 및 자동차의 대수가 60대 이상 넘어가면 에러를 띄우는 부분의 문제가 있어 수정했습니다.

jin123457 committed:
Remove : Race 테스트 코드 삭제

[삭제된 파일]
Race.test.js : Race 클래스는 섞여있는 부분이 많은 것 같아 다른 클래스에서 테스트 코드를 돌린 것이 나을 것 같아 삭제하였습니다.

jin123457 committed:
Docs : 예외 처리사항 목록 현황 반영 및 리팩토리 요구사항 현황 반영

[수정된 파일]
README.md : 예외 처리사항 목록 현황 반영 및 리팩토리 요구사항 현황 반영하였습니다.

jin123457 committed:
Rename : 파일 위치 이동 및 호출 경로 수정

[수정된 파일]
validation.test.js : 폴더 이동에 따른 호출 경로를 수정을 하였습니다.
- validation.js

Race.js : 폴더 이동에 따른 호출 경로를 수정을 하였습니다.
- random.js

Input.js :  폴더 이동에 따른 호출 경로를 수정을 하였습니다.
- readLine.js

Output.js :  폴더 이동에 따른 호출 경로를 수정을 하였습니다.
- positionToStr.js

RaceSetup.js :  폴더 이동에 따른 호출 경로를 수정을 하였습니다.
- tryInput.js
- validation.js

tryInput.js :  폴더 이동 및 폴더 이동에 따른 호출 경로를 수정을 하였습니다. (view/ -> utils/)
- errorHandler.js

errorHandler.js : 파일이 이동되었습니다. (view/ -> utils/)
positionToStr.js : 파일이 이동되었습니다. (domain/ -> utils/)
random.js : 파일이 이동되었습니다. (domain/ -> utils/)
readLine.js : 파일이 이동되었습니다. (view/ -> utils/)
validation.js : 파일이 이동되었습니다. (domain/ -> validation/)

jin123457 committed:
Chore : 경주 시작전 초기화 하는 변수의 네이밍 수정

[수정된 파일]
Race.js : raceSetup -> initializeRace로 좀 더 명확하게 수정하였습니다.

JeongBin0227 commented:
> controller 역할 분리

-> 개인적으로 너무 좋은거 같아요 코드도 깔끔해졌고, 단일 책임 원칙도 좋네요!

JeongBin0227 commented:
> Car의 moveForward 함수

-> 좋은 방향으로 리팩토링 됐다고 생각합니다!  전 오히려 단익 책임을 더 잘 지켰다고 생각합니다
더 자세히 적으면, 이전에는 
Car가 이동할 조건까지 알고 있었습니다 하지만 자동차 객체는 "이동 가능 조건(랜덤 숫자 >= 기준 값)"까지 알 필요가 없었고, 자동차는 그냥 "움직인다(move)"는 행위만 수행하면 됐다고 생각해요 그런점에서 차는 명확히 움직이는 역할만 하니까 더 좋았고 비즈니스 로직이 Race에 집중되다 보니 유지보수 측면에서도 룰이 변경되면 race 만 수정하면 되니까 더 좋네요!

JeongBin0227 commented on src/domain/Race.js:
```diff
@@ -0,0 +1,43 @@
+class Race {
+  async play() {
+    const initializeRace = new RaceSetup();
+    await initializeRace.carList();
+    await initializeRace.raceCount();
```
지금은 carList 와 raceCount 밖에 없는데 나중에 더 추가될것을 생각해서 initializeRace 로 묶어도 깔끔할거 같네요!

```
async play() {
  const raceSetup = new RaceSetup();
  await raceSetup.initializeRace();
  this.start(raceSetup.getCarList(), raceSetup.getRaceCount());
}
```
이런식이 될거 같아요

jin123457 replied:
확장성을 고려를 못했네요ㅠ
확실히 케빈이 추천해주신 방법으로 하면 나중에 추가 될 때 RaceSetup만 수정하면 되서 훨씬 깔끔해질 것 같아요!
피드백에 반영하겠습니다. 감사합니다.😊

JeongBin0227 commented on src/domain/Winners.js:
```diff
@@ -0,0 +1,17 @@
+  }
+}
+
+export default Winners;
```
이건 그냥 제 고민거리인데, 저는 이런 class 형식의 코드를 짤때 가끔씩 determine()이 실행되지 않은 상태에서 getNames()가 호출될 경우는 없을까? 라는 고민을 하곤 합니다. 혼자 개발을 하는거라면 상관 없겠지만 여러면 개발할때나 런타임에서 어떠한 이유로 determine()이 실행되지 않은 상태에서 getNames()가 호출되는 경우가 있지 않을까? 생각이 들어요

그래서 저는 그러면 에러가 발생하니 항상 
```
getNames() {
  return this.#names ? this.#names.join(', ') : 'No winners';
}
```

이런식으로 처리를 하는데 이런방식에 대해서 어떻게 생각하세요?

jin123457 replied:
케빈 정말 좋은 생각 같습니다.
훨씬 더 안정성있는 코드가 될 것 같아요!
``` js
getNames() {
  if(!this.#names) {
    return  'No winners';
  }

  return this.#names.join(', ');
}
```
프리코스 진행 중에 삼항연산자에 대해 사용하지 말라고 글이 있어서 삼항연산자를 제외한 이렇게 하는건 케빈은 어떠신가요?

JeongBin0227 replied:
아 그렇다면 저는 저런 코드도 좋다고 생각해요~
오히려 조건이 더 복잡해진다면 if문이 가독성이나 유지보수 측면에서 더 좋을 수도 있을거 같아요

JeongBin0227 commented on src/domain/Car.js:
```diff
@@ -1,4 +1,4 @@
+import { MOVE_UNIT } from '../constant/constant.js';
 
 class Car {
   #name;
```
👍  private 필드로 하셨군요

JeongBin0227 commented on src/validation/validation.js:
```diff
@@ -20,13 +21,23 @@ const validateLength = (names) => {
 };
 
+const validateNameOverlap = (names) => {
+  if (names.length !== new Set(names).size) {
```
Set()을 활용해서 중복 검사를 하셨군요 👍

JeongBin0227 commented:
> domain 로직과 UI 로직이 잘 구분되어 있나요?

-> 현재 domain/ 폴더에 핵심 비즈니스 로직 (Race.js, RaceSetup.js, Winners.js, Car.js)이 위치하기도 했고, view/ 폴더에는 UI 관련 코드 (Input.js, Output.js)가 따로 분리되어 있어서 잘한거 같아요. 하지만 Race.js가 직접 OutputView를 참조하는 방식이긴 하지만, 이 정도는 게임 실행 흐름을 담당하는 컨트롤러 역할이라 괜찮다고 생각합니다.

굳이 개선을 해보자면 UI 로직을 완전히 분리하기 위해 Race.js에서 OutputView 호출을 하지 않고, Race는 우승자 리스트를 반환하고, UI는 이를 출력하는 역할만 수행하도록 수정할 수 있을거 같아요

JeongBin0227 commented:
> utils 폴더가 적절하게 분리되었나요?

-> 현재 utils/ 폴더에 있는 random.js, tryInput.js, readLine.js 등은 도메인과 직접적인 연관이 없는 범용적인 기능이라고 생각해서 잘하신거 같아요!

JeongBin0227 commented:
> domain/이 view/를 의존하지 않도록 잘 구성되었나요?

-> domain/ 폴더 내 (Race.js, RaceSetup.js, Winners.js, Car.js)는 view/를 직접 참조하지 않습니다. 하지만 Race.js에서 OutputView를 참조하지만, 이는 비즈니스 로직이 아니라 컨트롤러 역할을 해서 괜찮다고 생각해요. 만약 추후 UI가 변경될 가능성이 있다면, Race.js에서 OutputView를 분리하는 방법도 고려해볼수 있을거 같아요

JeongBin0227 commented:
> PR 메시지 관련 피드백

-> 지금처럼 고민포인트를 적어주시는것도 좋고, 세오가 코드를 작성하면서 든 생각들, 이렇게 선택한 이유등이 같이 있으면 더 좋을거 같아요
PR 메시지에 남겨도 좋고 제가 코멘트 남기는것처럼 코드 중간중간 남겨도 좋을거 같아요

JeongBin0227 commented:
> 개발할 때 가장 설레고 재미있는 순간?

-> 저는 가장 크게는 에러를 해결했을때 설레고 재밌는거 같아요. 또 제가 잘 모르는 도메인에 관련한 개발을 할때 무척 설렙니다!

JeongBin0227 commented on src/domain/Race.js:
```diff
@@ -0,0 +1,43 @@
+    OutputView.printBeforeResult();
+    for (let i = 0; i < raceCount; i++) {
+      this.progressRace(carList);
+      console.log('');
```
nit) 사소하지만, 이렇게 줄바꿈이 필요할때는 OutputView.printEmptyLine() 이렇게 빼도 좋더라구요. 왜냐하면 리팩토링 하면서 저도 모르게 불필요한 console 인지 알고 지우는 경우가 종종 있어요

jin123457 replied:
저도 console.log를 안쓰고 어떻게 하는 방법이 있을까 궁금했는데 케빈의 방법 너무 좋은 것 같습니다.
코드를 볼 때 통일성도 있어보여 너무 좋은 것 같습니다.ㅎㅎ
감사합니다!

JeongBin0227 commented on src/domain/Race.js:
```diff
@@ -0,0 +1,43 @@
+
+  progressRace(carList) {
+    carList.forEach((car) => {
+      const moveCondition = randomNumber();
```
moveCondition은 마치 불리언 값을 가질 것처럼 보여서 randomMoveValue 도 괜찮을거 같아요~

jin123457 replied:
저는 moveCondition에서 randomNumber함수를 호출하니 숫자를 유추할 수 있다고 생각했었는데 케빈의 생각을 듣고 다시 한 번 보니 정말 그렇게 다시 보이네요. 항상 네이밍이 너무 어려운 것 같습니다ㅠ
감사합니다 😊

JeongBin0227 reviewed (changes requested):
안녕하세요~ 세오! 빠르게 2단계를 진행해주셔서 감사합니다 :) 

전체적으로 코드가 많이 깔끔해진거 같아요! 읽는데 아주 편했습니다 💯 

충분히 머지될만한 코드지만 코멘트나 질문에 대한 답변을 남겨서 한번 읽어보시라구 request changes 로 드립니다~ 다 확인이 되면 다시 리뷰 주세요~

2단계 미션하느라 고생많으셨어요 :)

jin123457 commented:
> -> do

jin123457 committed:
Refactor : 자동차와 경주 횟수 입력 함수 통합하여 호출

[수정된 파일]
Race.js : raceSetup 객체에서 사용자에게 입력받는 자동차 리스트, 경주 횟수를 따로따로 호출하는 것이 아닌 한 번에 통합하여 호출하는 방식으로 수정하였습니다.

jin123457 committed:
Fix : RaceSetup에서 initializeRace 함수 생성

[수정된 파일]
RaceSetup.js : Race 클래스에서 initializeRace를 호출하는데 RaceSetup 클래스에 해당 함수를 생성하지 않아 생성하였습니다.

jin123457 committed:
Refactor : winner 클래스의 getNames 함수 방어 코드 생성

[수정된 파일]
Winners.js : determine함수를 호출 전에 getNames 함수를 호출 할 때 에러가 나올 수 있는 상황을 대비해 방어 코드 생성하였습니다.
output.js : 우승자가 없을 경우 표현하는 문구를 추가했습니다.

jin123457 committed:
Chore : 공백 OutputView에 변수 추가 및 변수 네이밍 변경

JeongBin0227 reviewed (approved):
안녕하세요 세오~

제가 의도한대로 잘 반영해주셔서 감사합니다! 또 무작정 반영이 아닌 반영하는 과정에서 세오의 생각도 알려주셔서 너무 좋았습니다.

이제 충분히 머지할만큼 좋은 코드여서 이번 미션은 여기에서 머지하도록 하겠습니다
미션하느라 고생많으셨어요~
